internal/statesync/statesync.go

```diff
@@ -0,0 +1,316 @@
+package statesync
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/kwilteam/kwil-db/core/log"
+
+	rpchttp "github.com/cometbft/cometbft/rpc/client/http"
+)
+
+const (
+	ABCISnapshotQueryPath = "/snapshot/height"
+)
+
+// StateSyncer is responsible for initializing the database state from the
+// snapshots received from the network. It validates the snapshots against
+// the trusted snapshot providers.
+// The snapshot used to initialize are discarded after the database is restored.
+// Snapshot store if enabled, only stores snapshots produced by the node itself.
+
+type StateSyncer struct {
+	// statesync configuration
+	dbConfig *DBConfig
+
+	// directory to store snapshots and chunks
+	// same as the snapshot store directory
+	// as we allow to reuse the received snapshots
+	snapshotsDir string
+
+	// trusted snapshot providers for verification - cometbft rfc servers
+	snapshotProviders []*rpchttp.HTTP
+
+	// State syncer state
+	inProgress bool
+	snapshot   *Snapshot
+	chunks     map[uint32]bool // Chunks received till now
+	rcvdChunks uint32          // Number of chunks received till now
+
+	// Logger
+	log log.Logger
+}
+
+// NewStateSyncer will initialize the state syncer that enables the node to
+// receive and validate snapshots from the network and initialize the database state.
+// It takes the database configuration, snapshot directory, and the trusted snapshot providers.
+// Trusted snapshot providers are special nodes in the network trusted by the nodes and
+// have snashopt creation enabled. These nodes are responsible for creating and validating snapshots.
+func NewStateSyncer(ctx context.Context, cfg *DBConfig, snapshotDir string, providers []string, logger log.Logger) *StateSyncer {
+
+	ss := &StateSyncer{
+		dbConfig:     cfg,
+		snapshotsDir: snapshotDir,
+		log:          logger,
+	}
+
+	for _, s := range providers {
+		clt, err := ChainRPCClient(s)
+		if err != nil {
+			logger.Error("Failed to create rpc client", log.String("server", s), log.Error(err))
+			return nil
+		}
+		ss.snapshotProviders = append(ss.snapshotProviders, clt)
+	}
+
+	// Ensure that the snasphot directory exists and is empty
+	if err := os.RemoveAll(snapshotDir); err != nil {
```

Comment:
I didn't pick up on this before.  Why are we deleting any existing snapshots on start?  Doesn't that hurt the overall snapshot availability for the network as a whole?

Reply:
Yeah, I decided not use these snapshots used for statesync. Especially because this node might not have snapshots enabled and it makes management of snapshot store easy. also this directory is different from the snapshot direction u use if snapshot creation is enabled. This directory just stores chunks received during statesync and deletes them after restore. If incase snapshot creation is enabled, it would make its own chunks at later block heights.

Reply:
> this directory is different from the snapshot direction u use if snapshot creation is enabled. This directory just stores chunks received during statesync and deletes them after restore. If incase snapshot creation is enabled, it would make its own chunks at later block heights.

ahhhhhhhh